src/core/Akka.Streams/Dsl/Restart.cs

```diff
@@ -0,0 +1,490 @@
+            TimeSpan minBackoff,
+            TimeSpan maxBackoff,
+            double randomFactor)
+        {
+            FlowFactory = flowFactory;
+            MinBackoff = minBackoff;
+            MaxBackoff = maxBackoff;
+            RandomFactor = randomFactor;
+            Shape = new FlowShape<TIn, TOut>(In, Out);
+        }
+
+        public Inlet<TIn> In { get; } = new Inlet<TIn>("RestartWithBackoffFlow.in");
+        public Outlet<TOut> Out { get; } = new Outlet<TOut>("RestartWithBackoffFlow.out");
+
+        public override FlowShape<TIn, TOut> Shape { get; }
+
+        protected override GraphStageLogic CreateLogic(Attributes inheritedAttributes)
+        {
+            return new Logic(this, "Flow");
+        }
+
+        private class Logic : RestartWithBackoffLogic<FlowShape<TIn, TOut>>
+        {
+            private readonly RestartWithBackoffFlow<TIn, TOut, TMat> _stage;
+            private Tuple<SubSourceOutlet<TIn>, SubSinkInlet<TOut>> activeOutIn = null;
+
+            public Logic(RestartWithBackoffFlow<TIn, TOut, TMat> stage, string name)
+                : base(name, stage.Shape, stage.MinBackoff, stage.MaxBackoff, stage.RandomFactor)
+            {
+                _stage = stage;
+                Backoff();
+            }
+
+            protected override void StartGraph()
+            {
+                var sourceOut = CreateSubOutlet(_stage.In);
+                var sinkIn = CreateSubInlet(_stage.Out);
+                Source.FromGraph(sourceOut.Source).Via(_stage.FlowFactory()).RunWith(sinkIn.Sink, SubFusingMaterializer);
+                if (IsAvailable(_stage.Out))
+                {
+                    sinkIn.Pull();
+                }
+                activeOutIn = Tuple.Create(sourceOut, sinkIn);
+            }
+
+            protected override void Backoff()
+            {
+                SetHandler(_stage.In, () =>
+                {
+                    // do nothing
+                });
+                SetHandler(_stage.Out, () =>
+                {
+                    // do nothing
+                });
+
+                // We need to ensure that the other end of the sub flow is also completed, so that we don't
+                // receive any callbacks from it.
+                if (activeOutIn != null)
+                {
+                    var sourceOut = activeOutIn.Item1;
+                    var sinkIn = activeOutIn.Item2;
+                    if (!sourceOut.IsClosed)
+                    {
+                        sourceOut.Complete();
+                    }
+                    if (!sinkIn.IsClosed)
+                    {
+                        sinkIn.Cancel();
+                    }
+                    activeOutIn = null;
+                }
+            }
+        }
+    }
+
+    /// <summary>
+    /// Shared logic for all restart with backoff logics.
+    /// </summary>
+    internal abstract class RestartWithBackoffLogic<S> : TimerGraphStageLogic where S : Shape
```

Comment:
You could implement IInAndOutHandler and use SetHandler(Out/In, this) to save a few allocations, we’ve done that in many other stages.

Reply:
I'm working on microbenchmark set for Akka.Streams, and I will try to reduce some allocations in the future. I don;t like to do some unobvious performance improvements without any tests.